### ISSUE TITLE
about improvement over previous methods

### ISSUE THREAD

**Issue by lyk412** [guochengqian/Magic123#27]
Hi! Thanks for your awesome work! 
I noticed that:
1.  the `clarity` of your 3D reconstruction is very high, which shows a significant improvement compared to the previous method.  I really want to know the `improvements you made to the code`(stable-dreamfusion codebase)? Thank you very much!
2. In the first few epochs of training, the results of front perspective are already very good. The examples of epoch 3:
![magic123_metal_dragon_statue_nerf_coarse_ep0003_lambertian](https://github.com/guochengqian/Magic123/assets/65889678/9f8a7473-6ef2-4bb3-b64a-3088fd960184)

The results of `previous method`:
<img width="297" alt="stable-dreamfusion" src="https://github.com/guochengqian/Magic123/assets/65889678/b36c3873-683b-4036-877e-13bde4aafa0a">
The results of `magic123`:
<img width="268" alt="magic123" src="https://github.com/guochengqian/Magic123/assets/65889678/24b21d7e-4396-4e55-af96-46a5e6e3629b">

**Comment by guochengqian** [comment]
Could you clarify what do you mean by previous method?

**Comment by lyk412** [comment]
Thank you very much for your reply! 
Sorry, I didn't explain it clearly. I implemented your strategy on the codebase of `stable-dreamfusion`, providing 3D and 2D guidance at a fixed scale. But I found that the clarity of `the first stage` results is not good(fig.1), and your first stage result(fig2, obtained using `magic123 codebase`) is closer to the original image. 
So, I'm curious about what improvements have been made to the codebase of stabledramfusion?
`fig.1`
<img width="436" alt="dragon_stabledf" src="https://github.com/guochengqian/Magic123/assets/65889678/318f2563-4020-42c1-9918-3b1fe6691431">
`fig2`
<img width="284" alt="dragon_magic123" src="https://github.com/guochengqian/Magic123/assets/65889678/d50c7cba-71d5-4c6f-bfa7-5a2035d87aa2">

**Comment by guochengqian** [comment]
yeah. The changes is in the gradient of the SDS loss. They use sum, we use mean.

**Comment by guochengqian** [comment]
please check here, we comment this line: https://github.com/guochengqian/Magic123/blob/7f051875e48988dc624b2bd8425aea370c093588/guidance/sd_utils.py#L376

and use 
```
        # loss = SpecifyGradient.apply(latents, grad)
        # loss = loss.abs().mean().detach()
        latents.backward(gradient=grad, retain_graph=True)
```
